openage/convert/slp.pyx

```diff
@@ -500,59 +621,384 @@ cdef class SLPFrame:
         # end of row reached, return the created pixel array.
         return
 
-    cdef inline uint8_t get_byte_at(self, Py_ssize_t offset):
-        """
-        Fetch a byte from the slp.
-        """
-        return self.data_raw[offset]
 
-    cdef inline cmd_pack cmd_or_next(self, uint8_t cmd,
-                                     uint8_t n, Py_ssize_t pos):
+cdef class SLPMainFrameDE(SLPFrame):
+    """
+    SLPFrame for the main graphics sprite since SLP version 3.0.
+    """
+
+    def __init__(self, frame_info, data):
+        super().__init__(frame_info, data)
+
+    cdef process_drawing_cmds(self, vector[pixel] &row_data,
+                              Py_ssize_t rowid,
+                              Py_ssize_t first_cmd_offset,
+                              size_t expected_size):
         """
-        to save memory, the draw amount may be encoded into
-        the drawing command itself in the upper n bits.
+        create palette indices (colors) for the drawing commands
+        found for this row in the SLP frame.
         """
+        # position in the data blob, we start at the first command of this row
+        cdef Py_ssize_t dpos = first_cmd_offset
 
-        cdef uint8_t packed_in_cmd = cmd >> n
+        # is the end of the current row reached?
+        cdef bool eor = False
 
-        if packed_in_cmd != 0:
-            return cmd_pack(packed_in_cmd, pos)
+        cdef uint8_t cmd
+        cdef uint8_t nextbyte
+        cdef uint8_t lower_nibble
+        cdef uint8_t higher_nibble
+        cdef uint8_t lowest_crumb
+        cdef cmd_pack cpack
+        cdef int pixel_count
 
-        else:
-            pos += 1
-            return cmd_pack(self.get_byte_at(pos), pos)
+        # work through commands till end of row.
+        while not eor:
+            if row_data.size() > expected_size:
+                raise Exception(
+                    "Only %d pixels should be drawn in row %d, "
+                    "but we have %d already!" % (
+                        expected_size, rowid, row_data.size()
+                    )
+                )
 
-    def get_picture_data(self, palette, player_number=0):
-        """
-        Convert the palette index matrix to a colored image.
-        """
-        return determine_rgba_matrix(self.pcolor, palette, player_number)
+            # fetch drawing instruction
+            cmd = self.get_byte_at(dpos)
```

Comment:
we have to make sure this is properly resolved to a c++-only call. we can see that in the cythonized `html` file. otherwise this will really slow down the conversion.

Reply:
Wouldn't the function declaration with `cdef` ensure that?